### ISSUE TITLE
Cabal prefers undeclared module in project root

### ISSUE THREAD

**Issue by fendor** [haskell/cabal#7159]
**Describe the bug**
`cabal` prefers modules in the project root directory over modules with the same name defined in the respective source modules.
It ignores that the module it uses to compile is not listed in the .cabal file directly.

**To Reproduce**

```
git clone https://github.com/fendor/stunning-disco-cabal-bug.git
cd stunning-disco-cabal-bug
cabal v2-build
cabal v2-exec cabal-bug-example
# > Output is: 3
```

**Expected behavior**

The output should be:
```
Output is: 5
```

**System information**
 - NixOS
 - `ghc`: 8.8.4 
 - `cabal`: `3.2.0.0` and `3.4.0.0` (built from source)

**Comment by phadej** [comment]
This is how GHC works. `Cabal` tells the source directories and the module names to GHC, which "finds" the actual files then. `Cabal` cannot do better, if GHC doesn't have a control mechanism,

**Comment by phadej** [comment]
```
/opt/ghc/bin/ghc-8.8.4 --make -fbuilding-cabal-package -O -static -outputdir /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -odir /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -hidir /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -stubdir /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -i -i/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -itest -i/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/autogen -i/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/global-autogen -I/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/autogen -I/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/global-autogen -I/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test-tmp -optP-include -optP/codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/autogen/cabal_macros.h -hide-all-packages -Wmissing-home-modules -no-user-package-db -package-db /cabal/store/ghc-8.8.4/package.db -package-db /codetmp/stunning-disco-cabal-bug/dist-newstyle/packagedb/ghc-8.8.4 -package-db /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/package.conf.inplace -package-id base-4.13.0.0 -package-id cabal-bug-example-0.1.0.0-inplace -XHaskell2010 test/MyLibTest.hs -o /codetmp/stunning-disco-cabal-bug/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-bug-example-0.1.0.0/t/cabal-bug-example-test/build/cabal-bug-example-test/cabal-bug-example-test -hide-all-packages
```

there is CWD `-i` in there... hmm. I think it always was there, but I haven't ever noticed myself

**Comment by phadej** [comment]
Try to make this `True` to be `null (hsSourceDirs bi)`. How much stuff on Hackage breaks if that is changed. Hopefully so little we can change that, e.g. none of current Stackage.

https://github.com/haskell/cabal/blob/8f82032b21fa36e1a50f780f942fd3494087a54c/Cabal/src/Distribution/Simple/GHC/Internal.hs#L406

**Comment by fendor** [comment]
I *thought* `-i` was supposed to do the right thing here and avoid that `GHC` automatically adds `-i.` to every invocation?

**Comment by phadej** [comment]
Hmm. Wait. True.

Now I can reproduce this with a simple invocation of GHC:

```
% ghc -fforce-recomp -O -i -isrc MyLib
[1 of 1] Compiling MyLib            ( MyLib.hs, MyLib.o )
```

Note, it's `MyLib.hs`, not `src/MyLib.hs`.

**Comment by phadej** [comment]
https://downloads.haskell.org/ghc/latest/docs/html/users_guide/flags.html#finding-imports says that `-i` is finding imports. Maybe not finding the compilation targets?

**Comment by phadej** [comment]
Compare:
```
ghc -fforce-recomp -O -i -isrc test/MyLibTest.hs 
[1 of 2] Compiling MyLib            ( src/MyLib.hs, src/MyLib.o )
[2 of 2] Compiling Main             ( test/MyLibTest.hs, test/MyLibTest.o )
Linking test/MyLibTest ...
```

Here, we *import* `MyLib`, and then GHC find it in right place.

GHC is inconsistent here. But whether that is "by design", I don't know. If GHC developers will answer that Cabal should be explicit, i.e. rather than module names pass actual files

```haskell
% ghc-8.10.2 -fforce-recomp -O -i src/MyLib.hs 
[1 of 1] Compiling MyLib            ( src/MyLib.hs, src/MyLib.o )
```

then we can compile without `-i<dir>` flags at all:

```haskell
% ghc -fforce-recomp -O -i src/MyLib.hs test/MyLibTest.hs
[1 of 2] Compiling MyLib            ( src/MyLib.hs, src/MyLib.o )
[2 of 2] Compiling Main             ( test/MyLibTest.hs, test/MyLibTest.o )
Linking test/MyLibTest ...
```

which we probably *should do*, as this will effectively turn off any module discovery GHC does (which I was mistaken this issue at first).

TL;DR: Both are sufficient alone, but maybe worth doing both still:
- Either fix GHC to use `-idirs` to find "entry point" modules (I have no idea how big)
- Pass explicit list-file to `ghc`, not module list. Then it won't be needed to pass `hs-source-dirs` as `-i` arguments. This is big internal change to Cabal, but I'd like to see it.

**Comment by phadej** [comment]
Opened a GHC issue https://gitlab.haskell.org/ghc/ghc/-/issues/18952

**Comment by phadej** [comment]
Closing. I don't think there's anything Cabal should do to address this issue.
